File: stasis-core/src/de/boxxit/statis/serializer/LocalDateSerializer.java
package de.boxxit.statis.serializer;

import com.esotericsoftware.kryo.Kryo;
import com.esotericsoftware.kryo.Serializer;
import com.esotericsoftware.kryo.io.Input;
import com.esotericsoftware.kryo.io.Output;
import org.joda.time.DateTimeZone;
import org.joda.time.LocalDate;

/**
 * User: Christian Fruth
 */
public class LocalDateSerializer extends Serializer<LocalDate>
{
	public LocalDateSerializer()
	{
	}

	@Override
	public void write(Kryo kryo, Output output, LocalDate date)
	{
		long millis = date.toDateTimeAtStartOfDay(DateTimeZone.UTC).getMillis();
		output.writeLong(millis, true);
	}

	@Override
	public LocalDate read(Kryo kryo, Input input, Class<LocalDate> type)
	{
		long millis = input.readLong(true);
		return new LocalDate(millis, DateTimeZone.UTC);
	}
}

File: stasis-ios/Stasis/LocalDateSerializer.m
//
//  LocalDateSerializer.m
//  Stasis
//
//  Created by Christian Fruth on 08.03.13.
//  Copyright (c) 2013 Boxx IT Solutions e.K. All rights reserved.
//

#import "LocalDateSerializer.h"
#import "../JodaTime/LocalDate.h"
#import "Kryo.h"

@implementation LocalDateSerializer

static NSTimeZone *gmtTimeZone;

__attribute__((constructor)) static void initialize()
{
	gmtTimeZone = [NSTimeZone timeZoneWithName:@"GMT"];
}

- (BOOL)acceptsNull
{
	return NO;
}

- (BOOL)isFinal:(Class)type
{
	return YES;
}

- (void)write:(Kryo *)kryo value:(id)value to:(KryoOutput *)output
{
	LocalDate *date = value;
	NSInteger gmtSec = [gmtTimeZone secondsFromGMTForDate:date.value];
	NSInteger locSec = [[NSTimeZone localTimeZone] secondsFromGMTForDate:date.value];
	NSTimeInterval secDelta = locSec - gmtSec;
	UInt64 millis = (date.value.timeIntervalSince1970 + secDelta) * 1000;

	[output writeULong:millis];
}

- (id)read:(Kryo *)kryo withClass:(Class)type from:(KryoInput *)input
{
	// Zeit in GMT in lokale Zeitzone wandeln
	UInt64 millis = [input readULong];
	NSTimeInterval secondsSince1970 = millis / 1000;
	NSDate *aDate = [NSDate dateWithTimeIntervalSince1970:secondsSince1970];
	
	NSInteger gmtSec = [gmtTimeZone secondsFromGMTForDate:aDate];
	NSInteger locSec = [[NSTimeZone localTimeZone] secondsFromGMTForDate:aDate];
	NSTimeInterval secDelta = gmtSec - locSec;
	
	NSDate *bDate = [[NSDate alloc] initWithTimeInterval:secDelta sinceDate:aDate];
	
	return [LocalDate dateFromDate:bDate];
}

- (NSString *)getClassName:(Class)type
{
	return @"org.joda.time.LocalDate";
}

@end
